File: lib/src/tools/device.dart
import 'dart:io';

import 'package:adil192_linux/src/tools/result_of_command.dart';

/// Information about the user's device
final class Device {
  static bool hasAmdCpu() => cpuInfo.contains('amd');
  static bool hasIntelCpu() => cpuInfo.contains('intel');

  static bool hasAmdGpu() => gpuInfo.toLowerCase().contains('amd');
  static bool hasIntelGpu() => gpuInfo.toLowerCase().contains('intel');
  static bool hasNvidiaGpu() => gpuInfo.toLowerCase().contains('nvidia');

  static final cpuInfo = File('/proc/cpuinfo').readAsStringSync().toLowerCase();

  // lspci | egrep -i "vga|display|3d"
  static final gpuInfo = resultOfCommandSync('lspci', ['-mm'])
      .split('\n')
      .where(
        (line) =>
            line.toLowerCase().contains('vga') ||
            line.toLowerCase().contains('display') ||
            line.toLowerCase().contains('3d'),
      )
      .join('\n');
}

File: src/tools/device.rs
use std::fs;

use cached::once;
use cmd_lib::run_fun;

#[once()]
fn get_gpu_info() -> String {
  run_fun!(lspci "-mm" | grep -E -i "vga|display|3d")
    .unwrap()
    .to_lowercase()
}
#[once()]
fn get_cpu_info() -> String {
  fs::read_to_string("/proc/cpuinfo").unwrap().to_lowercase()
}

pub struct Device;
impl Device {
  #[allow(unused)]
  pub fn has_amd_cpu() -> bool {
    get_cpu_info().contains("amd")
  }
  pub fn has_intel_cpu() -> bool {
    get_cpu_info().contains("intel")
  }

  pub fn has_amd_gpu() -> bool {
    get_gpu_info().contains("amd")
  }
  pub fn has_intel_gpu() -> bool {
    get_gpu_info().contains("intel")
  }
  pub fn has_nvidia_gpu() -> bool {
    get_gpu_info().contains("nvidia")
  }
}
